Clears stdout when mpv begins encoding piped output

hedgehog90 opened the pull request:
Fixes a problem whereby warnings or messages output prior to encoding
would appear at the beginning of the encoded output stream,
rendering it unreadable to any media decoder.

Read this before you submit this pull request:
https://github.com/mpv-player/mpv/blob/master/DOCS/contribute.md

Reading this link and following the rules will get your pull request reviewed
and merged faster. Nobody wants lazy pull requests.

hedgehog90 commented:
> > Fixes a problem whereby warnings or messages output prior to encoding would appear at the beginning of the encoded output stream,
> 
> Why not just add `fflush(stderr);` before the actual encoding starts?

Presume you mean `fflush(stdout)` ?

avih commented:
> I'll update it to this.

Don't. I don't _know_ that it's the best solution, because I rarely use encoding and I don't recall bumping into this issue, and I didn't actually try this solution. I only _think_ it should be a reasonable approach considering the issue you described.

You should understand the problem well and find a good solution (assuming it needs a solution), and understand why it is a good solution, and then explain it to me ;)

hedgehog90 commented:
OK, well it assuming it clears the stream then it's better than setting the write position to 0.

I'm just going through some old feature branches that I wrote for my fork about 6 months ago and noticed I hadn't submitted this particular one to the official repo.
I use encoding mode almost exclusively and so it is well tested.

avih commented:
> would appear at the beginning of the encoded output stream,

Actually, you said part of the encoded stream - not on screen. That sounds outright wrong to me, and being unfamiliar with the data flow of messages while encoding, I don't have an obvious approach to the solution.

But if you want to fix it, then you need to be able to explain what causes the problem originally and, why your solution is good.

hedgehog90 commented:
Hm, it might be possible that this issue was only with a previous build where warnings were written to stdout, as I can't recreate this in my now up-to-date master branch.

avih commented:
>  I'll update the pull request.

Make sure you mention an exact and minimal way to reproduce the issue (because I still don't know how to), preferably with `--no-config` and without dependency on config files, explain what's the root cause, and how your fix addresses it.

avih commented:
> I refer to #8608 to recreate the issue simply.

That issue doesn't have a log, so it's hard to tell what's going on.

Akemi commented:
> I still don't see any explanation for the root cause of this issue. Why are warnings going to stdout and not to stderr? maybe that's the issue we need to fix? Why is your fix the best solution here?

isn't that the historical standard behaviour of mpv? see https://github.com/mpv-player/mpv/issues/3305 andhttps://github.com/mpv-player/mpv/issues/3130 for example.

avih commented:
> isn't that the historical standard behaviour of mpv? see #3305 and #3130 for example.

Hard to tell what it refers to specifically. Status line going to stderr is not an issue here, but errors/warnings have a very strong tradition (outside of mpv) of going to stderr.

avih commented:
Anyway, if it's intended that warnings to to stdout, then I don't see how encoding to stdout can work at all.

Warnings can also happen randomly during the encoding, and the fix at this PR does not address this case as far as I can tell. It only (apparently) hacks the case where there are startup warnings.

avih commented:
> I'm pretty sure the only warnings that get sent to stdout (for whatever historical reasons) are during initialization when it's processing the conf file.

That's entirely possible. But at this case you need to know that for a fact, and be able to explain how/where it happens at the code, how your patch addresses it, and then actually explain it to me.

hedgehog90 commented:
OK, I've messed up and you know a lot more than me about streams, thank you for clarifying a few things for me.

I assumed `fflush` would clear anything already written to the stream, but it writes what's in the buffer to the stream. I already knew this term but for some reason I assumed it did something different to what I was used to. So if I understand correctly it is impossible to metaphorically hit the delete key on a stream if I'm using a pipe.

So I've now reverted to `fseek(stdout, 0, SEEK_SET);` and realised my next error.

I assumed with `mpv --of=matroska --o=- video.mkv >out.mkv` that mpv delivered the same goods as `mpv --of=matroska --o=- | ffmpeg -i - out.mkv`... which it does in a sense, but now I understand that a file stream can be seeked, but a piped stream cannot, and so we get different results.
I had assumed the former command was an effective test, but it wasn't.
When I first encountered this issue, as soon as I found the error in my conf I corrected it, so naturally I never encountered the issue again when I output to a pipe.
I've just tested the other command and you are correct to assume it wouldn't work. ffmpeg errors with `pipe:: Invalid data found when processing input`

So now I'm thinking why are messages sent to stderr in the first place, and could we possibly change that?

Alternatively a fix might be to check if the loaded conf contains errors && mpv is in encoding mode && the output is a pipe, then exit before encoding can start.

hedgehog90 commented:
OK, well considering I might be the only one to encounter this bug, then it's probably best to just put a pin in it and make a brief note of it in the documentation below the `--o` argument.

Just to re-emphasise, the exact cause is when you have `--o=-` (or `--o=pipe:0`), there is the potential for warnings after processing a bad conf file to prepend the encoded output stream. There might be other ways of triggering messages before initializing the encoding that I'm not aware of.
`--o=<anything-else>` cannot trigger this, it has to be output to stdout.

Dudemanguy commented:
Did some digging into this (and yes the issue still exists today). So actually mpv **always** prints to stdout. The only exceptions to this are the status line (stderr) and encoding mode. As far as I can tell, this is completely intentional and what wm4 wanted. In this line:

https://github.com/mpv-player/mpv/blob/8eb7a00fa1f091e1766997f3d972d6c5fe9ac018/common/msg.c#L309

`root->force_stderr` is *never* set to true by anything other than encoding mode which is done by this:

https://github.com/mpv-player/mpv/blob/8eb7a00fa1f091e1766997f3d972d6c5fe9ac018/common/msg.c#L716

That function is only called by `common/encode.c` with the purpose of changing the out of messages while in encoding mode to stderr in the sole case when you are outputting the encode to stdout. If there are messages that already exist in stdout before that is called, they will end up in the encode and mess up the output. So this commit is trying to change the right thing. `fflush` won't work though. You'll still end up with garbage data in the stream. The previously mentioned `fseek(stdout, 0, SEEK_SET)` works however and is logical.

Not sure if you're still interested in this, but I would suggest adding that in `mp_msg_force_stderr`, changing that function to a bool, doing error checking on the fseek call, and then failing completely if that errors out for any reason.

sfan5 commented:
Solved elsewhere.
